### PR TITLE
Add queue manipulating endpoints

### PR DESCRIPTION
**Pull request by ashotvdovychenko**

### CODE REVIEW

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 1]
```diff
@@ -48,6 +49,27 @@ public Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId) {
...
+      if (!currentIndex.equals(newIndex)) {
+        var startIndex = Math.min(currentIndex, newIndex);
+        var endIndex = Math.max(currentIndex, newIndex);
+        var increment = currentIndex < newIndex ? -1 : 1;
+
+        entryRepository.updateEntryIndexes(queueId, startIndex, endIndex, increment);
+        entry.setEntryIndex(newIndex);
+      }
+    });
+  }
```
Comment (vetterocki) [inline src/main/java/com/qunite/api/service/QueueServiceImpl.java:66]: I feel myself puzzled about opportunity to "jump forward" along the queue. Currenty, if I have, let`s say, list of entries like this:
Entry index: 0, EntryId(memberId=7, queueId=1)
Entry index: 1, EntryId(memberId=6, queueId=1)
Entry index: 2, EntryId(memberId=5, queueId=1)
Entry index: 3, EntryId(memberId=4, queueId=1)
Entry index: 4, EntryId(memberId=3, queueId=1)

I have an opportunity to change member`s (with id=3) entry index from 4 to, for example, 1:
AFTER
Entry index: 0, EntryId(memberId=7, queueId=1)
Entry index: 1, EntryId(memberId=3, queueId=1)
Entry index: 2, EntryId(memberId=6, queueId=1)
Entry index: 3, EntryId(memberId=5, queueId=1)
Entry index: 4, EntryId(memberId=4, queueId=1)

Honestly speaking, I don`t see scenarios for such a behaviour and this feature of method can violate users order in queue, isn`t it?

File: src/main/java/com/qunite/api/web/controller/QueueController.java [hunk 2]
```diff
@@ -76,6 +79,34 @@ public ResponseEntity<Integer> memberPosition(@PathVariable(value = "id") Long q
     return ResponseEntity.of(queueService.getMemberPositionInQueue(memberId, queueId));
   }
 
+  @PostMapping("/{id}/members/{member-id}")
+  @Operation(summary = "Enroll member to queue")
+  public ResponseEntity<Void> enrollMemberToQueue(@PathVariable(value = "id") Long queueId,
+                                                  @PathVariable(value = "member-id")
```
Comment (xd6d) [inline src/main/java/com/qunite/api/web/controller/QueueController.java:85]: @PathVariable could be simplified as in membersAmount(@PathVariable Long id) method above

File: src/main/java/com/qunite/api/data/EntryRepository.java [hunk 3]
```diff
@@ -4,13 +4,19 @@
...
 public interface EntryRepository extends JpaRepository<Entry, EntryId> {
   boolean existsByMemberId(Long id);
 
-  @Query("SELECT e.id FROM Entry e where e.queue.id = :queueId")
+  @Query("SELECT e.id FROM Entry e where e.queue.id = :queueId ORDER BY e.entryIndex")
   List<EntryId> findEntriesIdsByQueueId(Long queueId);
+
+  @Modifying
+  @Query(value = "UPDATE Entry e SET e.entryIndex = e.entryIndex + :increment "
+      + "WHERE e.queue.id = :queueId AND e.entryIndex BETWEEN :startIndex AND :endIndex ")
```
Comment (xd6d) [inline src/main/java/com/qunite/api/data/EntryRepository.java:20]: specify value implicitly

File: src/main/java/com/qunite/api/security/SecurityConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  private final JwtAuthorizationFilter authorizationFilter;
+
+  @Bean
+  public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+    http.csrf(AbstractHttpConfigurer::disable)
+        .sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+        .and()
+        .authorizeHttpRequests(requests -> requests
+            .requestMatchers("/auth/**").permitAll()
+            .requestMatchers("/swagger-ui.html").permitAll()
```
Comment (xd6d) [inline src/main/java/com/qunite/api/security/SecurityConfiguration.java:28]: 2 lines could be combined in 1
and add more swagger endpoints as
.requestMatchers("/auth/**", "/swagger-ui/**", "/v3/api-docs/**", "/swagger-ui.html").permitAll()

File: src/main/java/com/qunite/api/web/controller/QueueController.java [hunk 5]
```diff
@@ -70,9 +73,37 @@ public ResponseEntity<Integer> membersAmount(@PathVariable Long id) {
...
+  }
+
+  @PostMapping("/{id}/members/{member-id}")
+  @Operation(summary = "Enroll member to queue")
+  public ResponseEntity<Void> enrollMemberToQueue(@PathVariable Long id,
+                                                  @PathVariable(value = "member-id")
+                                                  Long memberId) {
+    queueService.enrollMemberToQueue(memberId, id);
+    return ResponseEntity.ok().build();
+  }
```
Comment (xd6d) [inline src/main/java/com/qunite/api/web/controller/QueueController.java:88]: It looks like the admin's endpoint to enroll any user to any queue. I believe we need another similar endpoint for user to enroll hisself.

File: src/test/java/com/qunite/api/service/QueueServiceTest.java [hunk 6]
```diff
@@ -82,4 +87,75 @@ void testGettingMemberPositionInQueue() {
...
+  void testChangeMemberPositionForward() {
+    var expectedEntryIdList = List.of(
+        new EntryId(6L, 1L),
+        new EntryId(5L, 1L),
+        new EntryId(7L, 1L),
+        new EntryId(4L, 1L),
+        new EntryId(3L, 1L));
+
+    queueService.changeMemberPositionInQueue(7L, 1L, 2);
+    var actualEntryIdList = entryRepository.findEntriesIdsByQueueId(1L);
```
Comment (Kivanval) [inline src/test/java/com/qunite/api/service/QueueServiceTest.java:102]: Put the repetition of constants in a variable

File: src/test/java/com/qunite/api/service/QueueServiceTest.java [hunk 7]
```diff
@@ -82,4 +87,75 @@ void testGettingMemberPositionInQueue() {
...
+    var expectedEntryIdList = List.of(
+        new EntryId(7L, 1L),
+        new EntryId(6L, 1L),
+        new EntryId(5L, 1L),
+        new EntryId(4L, 1L),
+        new EntryId(3L, 1L));
+
+    queueService.changeMemberPositionInQueue(5L, 1L, 2);
+    var actualEntryIdList = entryRepository.findEntriesIdsByQueueId(1L);
+
```
Comment (Kivanval) [inline src/test/java/com/qunite/api/service/QueueServiceTest.java:135]: Put the repetition of constants in a variable

**Review by xd6d — APPROVED** [review]

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 8]
```diff
@@ -51,12 +53,43 @@ public Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId) {
...
+  @Transactional(isolation = Isolation.REPEATABLE_READ)
+  public void changeMemberPositionInQueue(Long memberId, Long queueId,
+                                          Integer newIndex, String username) {
+    if (isUserQueueCreatorOrManagerByCredentials(queueId, username)) {
+      entryRepository.findById(new EntryId(memberId, queueId)).ifPresent(entry -> {
+        var currentIndex = entry.getEntryIndex();
+        if (!currentIndex.equals(newIndex)) {
+          var startIndex = Math.min(currentIndex, newIndex);
+          var endIndex = Math.max(currentIndex, newIndex);
+          var increment = currentIndex < newIndex ? -1 : 1;
```
Comment (eternityduck) [inline src/main/java/com/qunite/api/service/QueueServiceImpl.java:66]: I really don't understand why are you using `var` even with such easy types like `int`. I am supposed to read the line of code to catch why increment is `var`, but not an `int`

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 9]
```diff
@@ -51,12 +53,43 @@ public Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId) {
...
+    } else {
+      throw new AccessDeniedException("User is not a creator or manager");
+    }
+  }
+
+  @Override
+  @Transactional
+  public void deleteMemberFromQueue(Long memberId, Long queueId, String username) {
+    if (isUserQueueCreatorOrManagerByCredentials(queueId, username)) {
+      entryRepository.deleteById(new EntryId(memberId, queueId));
```
Comment (xd6d) [inline src/main/java/com/qunite/api/service/QueueServiceImpl.java:81]: Correct me if I am wrong: 
Lets say we have example queue:
Entry index: 0, EntryId(memberId=7, queueId=1)
Entry index: 1, EntryId(memberId=6, queueId=1)
Entry index: 2, EntryId(memberId=5, queueId=1)
Entry index: 3, EntryId(memberId=4, queueId=1)
Entry index: 4, EntryId(memberId=3, queueId=1).
We perform deleteMemberFromQueue(5, 1, "admin") and we get the following: 
Entry index: 0, EntryId(memberId=7, queueId=1)
Entry index: 1, EntryId(memberId=6, queueId=1)
Entry index: 3, EntryId(memberId=4, queueId=1)
Entry index: 4, EntryId(memberId=3, queueId=1).
So, entry index 2 is missing. And our queue now looks like this:
```
1.  member 7
2.  member 6
4.  member 4
5.  member 3
```

third place is missing. 

And my question is: Do we agree with such logic or not?

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 10]
```diff
@@ -51,12 +53,43 @@ public Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId) {
...
+      });
+    } else {
+      throw new AccessDeniedException("User is not a creator or manager");
+    }
+  }
+
+  @Override
+  @Transactional
+  public void deleteMemberFromQueue(Long memberId, Long queueId, String username) {
+    if (isUserQueueCreatorOrManagerByCredentials(queueId, username)) {
```
[the remark this answers is not in the document]
Comment (ashotvdovychenko) [reply src/main/java/com/qunite/api/service/QueueServiceImpl.java:85]: I`ve already created issue for this feature:
https://github.com/q-unite/qunite-api/issues/87

**Review by xd6d — APPROVED** [review]

File: pom.xml [hunk 11]
```diff
@@ -66,7 +67,11 @@
             <artifactId>spring-security-test</artifactId>
             <scope>test</scope>
         </dependency>
-
+        <dependency>
+            <groupId>one.util</groupId>
+            <artifactId>streamex</artifactId>
+            <version>${streamex.version}</version>
+        </dependency>
```
Comment (Kivanval) [inline pom.xml:75]: Remove space between dependencies

File: src/main/java/com/qunite/api/data/EntryRepository.java [hunk 12]
```diff
@@ -4,13 +4,24 @@
...
+
+  @Modifying
+  @Query("UPDATE Entry e SET e.entryIndex = e.entryIndex + :increment "
+      + "WHERE e.queue.id = :queueId AND e.entryIndex BETWEEN :startIndex AND :endIndex ")
+  void updateEntryIndexes(Long queueId, Integer startIndex, Integer endIndex, Integer increment);
+
+  @Modifying
+  @Query("UPDATE Entry e SET e.entryIndex = e.entryIndex - 1 "
+      + "WHERE e.queue.id = :queueId AND e.entryIndex > :startIndex")
+  void decrementEntryIndexes(Long queueId, Integer startIndex);
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/data/EntryRepository.java:26]: This one is special case of updateEntryIndexes, reuse it

File: src/main/java/com/qunite/api/service/QueueService.java [hunk 13]
```diff
@@ -12,11 +12,15 @@ public interface QueueService {
...
-  void enrollMemberToQueue(Long memberId, Long queueId);
+  void enrollMemberToQueue(String username, Long queueId);
 
   Optional<Integer> getMembersAmountInQueue(Long queueId);
 
   Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId);
 
+  void changeMemberPositionInQueue(Long memberId, Long queueId, Integer newIndex, String username);
+
+  void deleteMemberFromQueue(Long memberId, Long queueId, String username);
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/QueueService.java:23]: I don't undertand, why we need suffix Queue in QueueService

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 14]
```diff
@@ -51,12 +54,47 @@ public Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId) {
...
+  @Transactional(isolation = Isolation.REPEATABLE_READ)
+  public void deleteMemberFromQueue(Long memberId, Long queueId, String username) {
+    var entry = entryRepository.findById(new EntryId(memberId, queueId))
+        .orElseThrow(() -> new EntryNotFoundException(
+            "Could not find entry by memberId %s and queueId %s".formatted(memberId, queueId)));
+    if (isUserQueueCreatorOrManagerByCredentials(queueId, username)) {
+      entryRepository.deleteById(entry.getId());
+      entryRepository.decrementEntryIndexes(entry.getQueue().getId(), entry.getEntryIndex());
+    } else {
+      throw new AccessDeniedException("User is not a creator or manager");
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/QueueServiceImpl.java:89]: Perhaps it is better to prefer to throw out always our mistakes, then we will be sure that it is our mistake for sure

File: src/main/java/com/qunite/api/web/controller/QueueController.java [hunk 15]
```diff
@@ -70,9 +73,46 @@ public ResponseEntity<Integer> membersAmount(@PathVariable Long id) {
...
+
+  @PostMapping("/{id}/members")
+  @Operation(summary = "Enroll member to queue", responses = {
+      @ApiResponse(responseCode = "200"),
+      @ApiResponse(responseCode = "404", content = @Content())
+  })
+  public ResponseEntity<Void> enrollMemberToQueue(@PathVariable Long id,
+                                                  Principal principal) {
+    queueService.enrollMemberToQueue(principal.getName(), id);
+    return ResponseEntity.ok().build();
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/controller/QueueController.java:89]: When entering a queue, we can return a place in the queue

File: src/main/java/com/qunite/api/web/controller/QueueController.java [hunk 16]
```diff
@@ -70,9 +73,46 @@ public ResponseEntity<Integer> membersAmount(@PathVariable Long id) {
...
+    queueService.deleteMemberFromQueue(memberId, id, principal.getName());
+    return ResponseEntity.noContent().build();
+  }
+
+  @PatchMapping("/{id}/entries")
+  @Operation(summary = "Change member position in queue", responses = {
+      @ApiResponse(responseCode = "200"),
+      @ApiResponse(responseCode = "404", content = @Content())
+  })
+  public ResponseEntity<Void> changeMemberPositionInQueue(@PathVariable Long id,
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/controller/QueueController.java:110]: Also don't understand suffix Queue in QueueController

File: src/test/java/com/qunite/api/service/QueueServiceTest.java [hunk 17]
```diff
@@ -82,4 +92,152 @@ void testGettingMemberPositionInQueue() {
...
+        new EntryId(6L, queueId),
+        new EntryId(4L, queueId),
+        new EntryId(3L, queueId));
+
+    queueService.deleteMemberFromQueue(5L, queueId, "First");
+    var actualEntryIdList = entryRepository.findEntriesIdsByQueueId(queueId);
+
+    assertThat(expectedEntryIdList).isEqualTo(actualEntryIdList);
+    assertThat(actualEntryIdList).doesNotContainNull();
+  }
```
Comment (Kivanval) [inline src/test/java/com/qunite/api/service/QueueServiceTest.java:204]: Method does not test delete

File: src/main/java/com/qunite/api/web/controller/QueueController.java [hunk 18]
```diff
@@ -70,9 +73,46 @@ public ResponseEntity<Integer> membersAmount(@PathVariable Long id) {
...
+  })
+  public ResponseEntity<Void> deleteMemberFromQueue(@PathVariable Long id,
+                                                    @PathVariable(value = "member-id")
+                                                    Long memberId,
+                                                    Principal principal) {
+    queueService.deleteMemberFromQueue(memberId, id, principal.getName());
+    return ResponseEntity.noContent().build();
+  }
+
+  @PatchMapping("/{id}/entries")
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/controller/QueueController.java:105]: Why don't we use id path with the patch?

File: src/test/java/com/qunite/api/service/QueueServiceTest.java (hunk 17 above)
Comment (Kivanval) [inline src/test/java/com/qunite/api/service/QueueServiceTest.java:204]: Add test for cannot be deleted by a non-admin

File: src/main/java/com/qunite/api/service/QueueService.java (hunk 13 above)
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/QueueService.java:23]: Rename param "username" for something more specific

File: src/main/java/com/qunite/api/web/controller/QueueController.java (hunk 15 above)
Comment (ashotvdovychenko) [reply src/main/java/com/qunite/api/web/controller/QueueController.java:89]: I've created an issue for the refactor of this endpoint:
https://github.com/q-unite/qunite-api/issues/90

**Review by Kivanval — APPROVED** [review]

**Pull request merged**